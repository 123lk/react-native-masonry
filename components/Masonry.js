'use strict';

import { ListView, Image } from 'react-native';
import React, { Component } from 'react';
import Row from './Row';
import styles from '../styles/main';

export default class Masonry extends Component {
  static propTypes = {
    images: React.PropTypes.array,
  };

  static defaultProps = {
    images: []
  };

  constructor(props) {
    super(props);

    this.ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2})
    this.state = {
      dataSource: this.ds
    };

    Promise.all(__getImages(props.data))
      .then(fetchedImages => {
        this.state = {
          dataSource: this.ds.cloneWithRows(fetchedImages)
        };
      })
      .catch(console.log);

  }

  componentWillReceiveProps(nextProps) {
    Promise.all(__getImages(nextProps.data))
      .then(fetchedImages => {
        const images = __splitIntoColumns(fetchedImages);
        this.setState({
          dataSource: this.state.dataSource.cloneWithRows([...images])
        });
      })
      .catch(console.log);
  }

  render() {
    return (
       <ListView
        contentContainerStyle={ styles.masonry__container }
        dataSource={ this.state.dataSource }
        renderRow={ (data) => <Row {...data} /> }
      />
    )
  }
}


export function __getImages(images) {
  return images.map((image) => pGetImageSize(image.uri).then((dimensions) => {
    return {
      uri: image.uri,
      dimensions: {
        width: dimensions[0],
        height: dimensions[1]
      }
    };
  }));
}

export function __splitIntoColumns(data, nColumns = 2) {
  const dataSet = [];
  for (let i = 0; i < data.length; i++) {
    const index = i % nColumns;
    if (dataSet[index]) {
      dataSet[index].push(data[i]);
    } else {
      dataSet.push([data[i]]);
    }
  }
  return dataSet;
}

function pGetImageSize(uri) {
  return new Promise((resolve, reject) => {
      Image.getSize(uri, (width, height) => resolve([width, height]), (err) => reject(err));
  });
}
