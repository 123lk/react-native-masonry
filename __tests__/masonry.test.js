import Masonry from '../components/Masonry';
import { Text, View } from 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const brickSet = [
  {
    id: 1,
    uri: 'https://cat1.jpg'
  },
  {
    id: 2,
    routeId: 'cat-3',
    uri: 'http://test.com/cat3.jpg',
    onPress: (brick) => redirect(brick.routeId)
  },
  {
    id: 3,
    uri: 'http://test.com/cat2.jpg'
  },
  {
    id: 4,
    uri: 'http://test.com/cat3.jpg'
  }
];

test('MAIN: Render masonry correct', () => {
 const masonry = renderer.create(
  <Masonry
    bricks={brickSet}
   columns={3} />).toJSON();
  // First child should the scroll view
  const scrollView = masonry.children[0];
  expect(scrollView.type).toBe('RCTScrollView');

  // Crucial styles for the grid to work
  expect(scrollView.props.contentContainerStyle).toEqual(
    {
      flex: 1,
      justifyContent: 'space-between',
      flexDirection: 'row',
      width: '100%'
    });
});

// Ignore for now, until we can resolve jest issues
// test('SNAPSHOT: All functionality should match prev snapshot', () => {
//   const tree = renderer.create(
//       <Masonry
//     bricks={brickSet}
//     columns={3} />
//   ).toJSON();
//   expect(tree).toMatchSnapshot();
// });
