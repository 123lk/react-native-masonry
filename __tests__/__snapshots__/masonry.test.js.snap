// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT: All functionality should match prev snapshot 1`] = `
<View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
    dataSource={
      ListViewDataSource {
        "items": 0,
      }
    }
    renderRow={[Function]}
    renderScrollComponent={[Function]}
  >
    <View />
  </RCTScrollView>
</View>
`;
